crates/pallet-domains/src/staking.rs

```diff
@@ -977,40 +974,66 @@ pub(crate) fn do_unlock_funds<T: Config>(
     Withdrawals::<T>::try_mutate_exists(operator_id, nominator_id.clone(), |maybe_withdrawal| {
         let withdrawal = maybe_withdrawal.as_mut().ok_or(Error::MissingWithdrawal)?;
         do_convert_previous_epoch_withdrawal::<T>(operator_id, withdrawal)?;
-        let WithdrawalInBalance {
-            domain_id,
-            unlock_at_confirmed_domain_block_number,
-            amount_to_unlock,
-            storage_fee_refund,
-        } = withdrawal
-            .withdrawals
-            .pop_front()
-            .ok_or(Error::MissingWithdrawal)?;
+
+        ensure!(!withdrawal.withdrawals.is_empty(), Error::MissingWithdrawal);
 
         let latest_confirmed_block_number =
-            Pallet::<T>::latest_confirmed_domain_block_number(domain_id);
+            Pallet::<T>::latest_confirmed_domain_block_number(operator.current_domain_id);
+
+        let mut total_unlocked_amount: BalanceOf<T> = Zero::zero();
+        let mut total_storage_fee_refund: BalanceOf<T> = Zero::zero();
+        loop {
+            if withdrawal
+                .withdrawals
+                .front()
+                .map(|w| w.unlock_at_confirmed_domain_block_number > latest_confirmed_block_number)
+                .unwrap_or(true)
+            {
+                break;
+            }
+
+            let WithdrawalInBalance {
+                amount_to_unlock,
+                storage_fee_refund,
+                ..
+            } = withdrawal
+                .withdrawals
+                .pop_front()
+                .expect("Must not empty as checked above; qed");
+
+            // deduct the amount unlocked from total
+            withdrawal.total_withdrawal_amount = withdrawal
+                .total_withdrawal_amount
+                .checked_sub(&amount_to_unlock)
+                .ok_or(Error::BalanceUnderflow)?;
+
+            total_unlocked_amount = total_unlocked_amount
+                .checked_add(&amount_to_unlock)
+                .ok_or(Error::BalanceUnderflow)?;
+
+            total_storage_fee_refund = total_storage_fee_refund
+                .checked_add(&storage_fee_refund)
+                .ok_or(Error::BalanceUnderflow)?;
```

Comment:
This is actually overflow, not underflow. Though I don't see why would you want to use checked addition here given that subtraction succeeded above and for overflow to happen you'll need to exceed `u128`, which is not possible even with all tokens in existence.